lib/modules/manager/asdf/extract.ts

```diff
@@ -0,0 +1,58 @@
+import is from '@sindresorhus/is';
+import { logger } from '../../../logger';
+import { isSkipComment } from '../../../util/ignore';
+import { regEx } from '../../../util/regex';
+import { GithubTagsDatasource } from '../../datasource/github-tags';
+import * as nodeVersioning from '../../versioning/node';
+import type { PackageDependency, PackageFile } from '../types';
+
+const upgradeableTooling: Record<
+  string,
+  Pick<
+    PackageDependency,
+    'depName' | 'datasource' | 'packageName' | 'versioning'
+  >
+> = {
+  nodejs: {
+    depName: 'node',
+    datasource: GithubTagsDatasource.id,
+    packageName: 'nodejs/node',
+    versioning: nodeVersioning.id,
+  },
+};
+
+export function extractPackageFile(content: string): PackageFile | null {
+  logger.trace('asdf.extractPackageFile()');
+
+  const regex = regEx(
+    /^(?<toolName>(\w+))\s+(?<version>[^\s#]+)(?:[^#]*(#(?<comment>.*)))?/gm
```

Comment:
This regex works with the tests across multi line files

Reply:
```suggestion
    /^(?<toolName>(\w+))\s+(?<version>[^\s#]+)(?:[^#]*)?(?:\s#(?<comment>.*))?/gm
```

your regex doesn't support this
```
nodejs 14.0.0 15.0.0
```

it also needs a space before comment

Reply:
https://regex101.com/r/4Rhc86/1
https://regex101.com/r/cq8Fg2/1

i think `nodejs 1.2.3#comment` is invalid

Reply:
https://regex101.com/r/uzp5RO/1
```suggestion
    /^(?<toolName>(\w+))\s+(?<version>[^\s#]+)(?:\s[^\s#]*)*(?:\s#(?<comment>.*))?$/gm
```

Reply:
> your regex doesn't support this
> nodejs 14.0.0 15.0.0

there is a working test for that scenario so I am not sure why you think that

> i think nodejs 1.2.3#comment is invalid

yup but the supplied regex allowed for that so I thought you knew something I did not here, tried to get a balance of both

if the entry is 

```
nodejs 10.10.0# comment
```

then is the expectation that we would send `10.10.0#` to renovate which would then reject it due to unsupported version?

Reply:
this is the test I am thinking

```
const invalidComments = [
        {
          entry: 'nodejs 16.16.0# comment without leading space',
          expect: '16.16.0#',
        },
        {
          entry: 'nodejs 16.16.0#comment with no spaces',
          expect: '16.16.0#comment',
        },
      ];

      describe.each(invalidComments)(
        'invalid comment placements are passed through to renovate',
        (data) => {
          it(`entry: '${data.entry}'`, () => {
            const res = extractPackageFile(data.entry);
            expect(res).toEqual({
              deps: [
                {
                  currentValue: data.expect,
                  datasource: 'github-tags',
                  depName: 'node',
                  packageName: 'nodejs/node',
                  versioning: 'node',
                },
              ],
            });
          });
        }
      );
```

Reply:
```suggestion
    /^(?<toolName>(\w+)) (?<version>[^\s#]+)(?: [^\s#]+)* *(?: #(?<comment>.*))?$/gm
```

this is the right regex.
- only spaces are allowed, so no `\s` which allows newlines
- only a single spaces is allowed as delimitter
- https://regex101.com/r/uzp5RO/2

Reply:
makes sense - I've updated